src/hotspot/share/cds/filemap.cpp

```diff
@@ -1069,117 +1075,185 @@ class FileHeaderHelper {
-    size_t n = os::read(_fd, base_name, (unsigned int)name_size);
-    if (n != name_size) {
-      log_info(cds)("Unable to read base archive name from archive");
-      FREE_C_HEAP_ARRAY(char, base_name);
-      return nullptr;
+    assert(_header != nullptr && _is_valid, "must be a valid archive file");
+    return _header;
+  }
+
+  const char* base_archive_name() {
+    assert(_header != nullptr && _is_valid, "must be a valid archive file");
+    return _base_archive_name;
+  }
+
+ private:
+  bool check_crc() {
+    if (VerifySharedSpaces) {
+      FileMapHeader* header = (FileMapHeader*)_header;
+      int actual_crc = header->compute_crc();
+      if (actual_crc != header->crc()) {
+        log_info(cds)("_crc expected: %d", header->crc());
+        log_info(cds)("       actual: %d", actual_crc);
+        FileMapInfo::fail_continue("Header checksum verification failed.");
+        return false;
+      }
     }
-    if (base_name[name_size - 1] != '\0' || strlen(base_name) != name_size - 1) {
-      log_info(cds)("Base archive name is damaged");
-      FREE_C_HEAP_ARRAY(char, base_name);
-      return nullptr;
+    return true;
+  }
+
+  bool check_and_init_base_archive_name() {
+    unsigned int name_offset = _header->_base_archive_name_offset;
+    unsigned int name_size   = _header->_base_archive_name_size;
+    unsigned int header_size = _header->_header_size;
+
+    if (name_offset + name_size < name_offset) {
+      FileMapInfo::fail_continue("base_archive_name/size overflow: " UINT32_FORMAT "/" UINT32_FORMAT,
+                                 name_offset, name_size);
+      return false;
     }
-    if (!os::file_exists(base_name)) {
-      log_info(cds)("Base archive %s does not exist", base_name);
-      FREE_C_HEAP_ARRAY(char, base_name);
-      return nullptr;
+    if (_header->_magic == CDS_ARCHIVE_MAGIC) {
+      if (name_offset != 0) {
+        FileMapInfo::fail_continue("static shared archive must have zero _base_archive_name_offset");
+        return false;
+      }
+      if (name_size != 0) {
+        FileMapInfo::fail_continue("static shared archive must have zero _base_archive_name_size");
+        return false;
+      }
+    } else {
+      assert(_header->_magic == CDS_DYNAMIC_ARCHIVE_MAGIC, "must be");
+      if ((name_size == 0 && name_offset != 0) ||
+          (name_size != 0 && name_offset == 0)) {
+        // If either is zero, both must be zero. This indicates that we are using the default base archive.
+        FileMapInfo::fail_continue("Invalid base_archive_name/size: " UINT32_FORMAT "/" UINT32_FORMAT,
+                                   name_offset, name_size);
+        return false;
+      }
+      if (name_size > 0) {
+        if (name_offset + name_size > header_size) {
+          FileMapInfo::fail_continue("Invalid base_archive_name/size (out of range): "
+                                     UINT32_FORMAT " + " UINT32_FORMAT " > " UINT32_FORMAT ,
```

Comment:
Similar to above, maybe change the message to:

"Invalid base_archive_name offset/size (out of range)..."